src/sqlfluff/core/parser/lexer.py

```diff
@@ -20,12 +20,63 @@
 from sqlfluff.core.templaters import TemplatedFile
 from sqlfluff.core.config import FluffConfig
 from sqlfluff.core.templaters.base import TemplatedFileSlice
-from sqlfluff.core.slice_helpers import is_zero_slice, offset_slice
+from sqlfluff.core.slice_helpers import is_zero_slice, offset_slice, to_tuple
 
 # Instantiate the lexer logger
 lexer_logger = logging.getLogger("sqlfluff.lexer")
 
 
+class BlockTracker:
+    """This is an object for keeping track of templating blocks.
+
+    Using the .enter() and .exit() methods on opening and closing
+    blocks, we can match up tags of the same level so that later
+    it's easier to treat them the same way in the linting engine.
+
+    In case looping means that we encounter the same block more
+    than once, we use cache uuids against their source location
+    so that if we try to re-enter the block again, it will get
+    the same uuid on the second pass.
+    """
+
+    _stack: List[UUID] = []
+    _map: Dict[Tuple[int, int], UUID] = {}
+
+    def enter(self, src_slice: slice) -> None:
+        """Add a block to the stack."""
+        key = to_tuple(src_slice)
+        uuid = self._map.get(key, None)
+
+        if not uuid:
+            uuid = uuid4()
```

Comment:
In the JinjaAnalyzer, I initially used UUIDs, then later I realized that a simple counter had advantages:
* Same results every time
* Easier to read in debugging output (e.g. "0000001" vs a long random string)

Worth considering here?

Reply:
Based on a quick investigation, I think you're right that we _could_ do that, but I don't know if it's actually _simpler_. Good shout however.

Reply:
Mainly useful for debugging. No meaningful difference from a technical point of view.